compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,111 @@
+use rustc_abi::Abi;
+use rustc_hir::CRATE_HIR_ID;
+use rustc_middle::ty::{self, Instance, InstanceKind, ParamEnv, Ty, TyCtxt};
+use rustc_session::lint::builtin::ABI_ERROR_DISABLED_VECTOR_TYPE;
+use rustc_span::def_id::DefId;
+use rustc_span::{Span, Symbol};
+use rustc_target::abi::call::{FnAbi, PassMode};
+
+use crate::errors::{AbiErrorDisabledVectorTypeCall, AbiErrorDisabledVectorTypeDef};
+
+// Represents the least-constraining feature that is required for vector types up to a certain size
+// to have their "proper" ABI.
+const X86_VECTOR_FEATURES: &'static [(u64, &'static str)] =
+    &[(128, "sse"), (256, "avx"), (512, "avx512f")];
+
+const AARCH64_VECTOR_FEATURES: &'static [(u64, &'static str)] = &[(128, "neon")];
+
+fn do_check_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    abi: &FnAbi<'tcx, Ty<'tcx>>,
+    target_feature_def: DefId,
+    emit_err: impl Fn(&'static str),
+) {
+    let feature_def = if tcx.sess.target.arch == "x86" || tcx.sess.target.arch == "x86_64" {
+        X86_VECTOR_FEATURES
+    } else if tcx.sess.target.arch == "aarch64" {
+        AARCH64_VECTOR_FEATURES
+    } else {
+        // FIXME: add support for non-tier1 architectures
+        return;
+    };
+    let codegen_attrs = tcx.codegen_fn_attrs(target_feature_def);
+    for arg_abi in abi.args.iter().chain(std::iter::once(&abi.ret)) {
+        let size = arg_abi.layout.size;
+        if matches!(arg_abi.layout.abi, Abi::Vector { .. })
+            && !matches!(arg_abi.mode, PassMode::Indirect { .. })
```

Comment:
This check isn't quite right for `PassMode::Cast`. In that case the `cast` field defines which registers will be used.

I don't know if we ever pass SIMD vectors via `Cast`. So the alternative would be an assertion. But we should make sure the logic doesn't silently accept bad calls.

Reply:
I added a check that _might_ be correct to my understanding of PassMode::Cast, but PTAL

Reply:
Yeah that looks good!